### PR TITLE
[batch-beta] Fix some bugs in file egress

### PR DESCRIPTION
**Pull request by matthchr**

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@matthchr__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (matthchr). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by matthchr** [comment]
CI is failing because I used `let` -- I am checking with Amar if this is intended or just an artifact of the way that they are running jshint

File: lib/commands/batch/fileegress/uploader.py [hunk 1]
```diff
@@ -88,7 +109,7 @@ def _extract_pathinfo(input_pattern):
     recursive = False
     fullpath = True
     for p in _pparts:
-        if p.startswith('*'):
+        if p.find('*') != -1:
```
Comment (annatisch) [inline lib/commands/batch/fileegress/uploader.py]: You can just use "if '*' in p:"

File: lib/commands/batch/batch.templateUtils._js [hunk 2]
```diff
@@ -168,33 +171,38 @@ batchTemplateUtils.processJobForOutputFiles = function(job, tasks) {
     var command = util.format('%s > $AZ_BATCH_JOB_PREP_DIR/uploadsetuplog.txt 2>&1', setupCommandLine);
     var resourceFiles = [];
 
-    // TODO: If we have any issues with thsi being hosted in GITHUB we'll have to move it elsewhere (local and then upload to their storage?)
+    let rootUrl = 'https://raw.githubusercontent.com/Azure/azure-xplat-cli/batch-beta';
```
Comment (annatisch) [inline lib/commands/batch/batch.templateUtils._js]: Maybe we could put this at the top to make it easier to find

File: lib/commands/batch/fileegress/configuration.py [hunk 3]
```diff
@@ -61,11 +73,15 @@ def __init__(self, container):
     def from_dict(dict):
```
Comment (annatisch) [inline lib/commands/batch/fileegress/configuration.py]: While it's probably a bit late - 'dict' is a keyword, so probably should be avoided as a parameter name.

File: lib/commands/batch/fileegress/test/perf.py [hunk 4]
```diff
@@ -16,8 +16,7 @@ def file_fixture(tmpdir, request):
     file_size_bytes = 1024*1024*file_size_mb
     test_end_to_end.create_files(
         file_count,
-        str(tmpdir),
-        'abc',
+        str(tmpdir / 'abc'),
```
Comment (annatisch) [inline lib/commands/batch/fileegress/test/perf.py]: This line confuses me... are these supposed to be joined?

File: lib/commands/batch/fileegress/test/test_end_to_end.py [hunk 5]
```diff
@@ -8,34 +8,50 @@
 except:
     import urlparse
     pass
-
+try:
+    import pathlib
+except:
+    import pathlib2 as pathlib
+    pass
```
Comment (annatisch) [inline lib/commands/batch/fileegress/test/test_end_to_end.py]: minor: pass isn't required here

File: lib/commands/batch/fileegress/batchfileuploader.py [hunk 6]
```diff
@@ -68,6 +68,16 @@ def load_specification_from_env(env):
...
 
 
+def load_specification_from_stdin():
+    # type: () -> configuration.Specification
+    """Loads a specification from stdin
+
+    :return: The specification object
+    """
+    spec = json.loads(sys.stdin.read())
+    return configuration.Specification.from_dict(spec)
```
Comment (annatisch) [inline lib/commands/batch/fileegress/batchfileuploader.py]: Do we need any confirmation that the stdin stream will be JSON compatible?
Or are you happy for that to simply be raised?

File: lib/commands/batch/fileegress/configuration.py [hunk 7]
```diff
@@ -15,6 +15,14 @@
 #
 
 import enum
+import json
+
+
+class SpecificationEncoder(json.JSONEncoder):
+    def default(self, o):
+        if isinstance(o, enum.Enum):
```
Comment (annatisch) [inline lib/commands/batch/fileegress/configuration.py]: Does this need to be compatible with 2.7 and 3.4? I think from memory isinstance(enum.Enum) had some issues in the 2.7 back port of the library... but I guess if this is working in your tests then it should be fine.

Comment (annatisch) [inline lib/commands/batch/fileegress/configuration.py]: You could potentially just do (assuming of course there's no other "value" attributes in your objects):
def default(self, o):
    try:
        return o.value
    except AttributeError:
        return o.__dict__

File: lib/commands/batch/fileegress/test/test_extractpathinfo.py [hunk 8]
```diff
@@ -1,7 +1,11 @@
 import os
 import util
-import pytest
 import uploader
+try:
+    import pathlib
+except:
+    import pathlib2 as pathlib
+    pass
```
Comment (annatisch) [inline lib/commands/batch/fileegress/test/test_extractpathinfo.py]: minor: Don't need pass again

File: lib/commands/batch/fileegress/configuration.py (hunk 3 above)
Comment (matthchr) [inline lib/commands/batch/fileegress/configuration.py]: Fixed

File: lib/commands/batch/batch.templateUtils._js (hunk 2 above)
Comment (matthchr) [inline lib/commands/batch/batch.templateUtils._js]: Done

File: lib/commands/batch/fileegress/uploader.py (hunk 1 above)
Comment (matthchr) [inline lib/commands/batch/fileegress/uploader.py]: Nice, fixed

File: lib/commands/batch/fileegress/batchfileuploader.py (hunk 6 above)
Comment (matthchr) [inline lib/commands/batch/fileegress/batchfileuploader.py]: I think it's okay to just throw in that case

File: lib/commands/batch/fileegress/test/test_end_to_end.py (hunk 5 above)
Comment (matthchr) [inline lib/commands/batch/fileegress/test/test_end_to_end.py]: Fixed

File: lib/commands/batch/fileegress/test/test_extractpathinfo.py (hunk 8 above)
Comment (matthchr) [inline lib/commands/batch/fileegress/test/test_extractpathinfo.py]: Fixed

File: lib/commands/batch/fileegress/test/perf.py (hunk 4 above)
Comment (annatisch) [inline lib/commands/batch/fileegress/test/perf.py]: Ah gotcha :)

**Comment by matthchr** [comment]
Xing suggested I merge this PR to batch-beta instead of batch-beta-dev since it is mostly bugfixes.

I think I will do that, so make any more comments you have but please don't merge this

**Pull request closed without merging**